velox/exec/HashBuild.cpp

```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
+       ++partition) {
+    const int numSpillInputs = numSpillInputs_[partition];
+    if (numSpillInputs == 0) {
+      continue;
+    }
+    VELOX_CHECK(spiller_->isSpilled(partition));
+    spillPartition(
+        partition, numSpillInputs, spillInputIndicesBuffers_[partition], input);
+  }
+  activeRows_.updateBounds();
+}
+
+void HashBuild::prepareInputIndicesBuffers(
+    vector_size_t numInput,
+    const SpillPartitionNumSet& spillPartitions) {
+  for (const auto& partition : spillPartitions) {
+    if (spillInputIndicesBuffers_[partition] == nullptr) {
+      spillInputIndicesBuffers_[partition] = allocateIndices(numInput, pool());
+    } else {
+      AlignedBuffer::reallocate<vector_size_t>(
+          &spillInputIndicesBuffers_[partition], numInput);
+    }
+  }
+  std::fill(numSpillInputs_.begin(), numSpillInputs_.end(), 0);
+}
+
+void HashBuild::computeSpillPartitions(const RowVectorPtr& input) {
+  if (hashes_.size() < activeRows_.size()) {
+    hashes_.resize(activeRows_.size());
+  }
+  const auto& hashers = table_->hashers();
+  for (auto i = 0; i < hashers.size(); ++i) {
+    auto& hasher = hashers[i];
+    if (hasher->channel() != kConstantChannel) {
+      hashers[i]->hash(activeRows_, i > 0, hashes_);
+    } else {
+      hashers[i]->hashPrecomputed(activeRows_, i > 0, hashes_);
+    }
+  }
+
+  spillPartitions_.resize(input->size());
+  for (auto i = 0; i < spillPartitions_.size(); ++i) {
+    spillPartitions_[i] = spiller_->hashBits().partition(hashes_[i]);
+  }
+}
+
+void HashBuild::spillPartition(
+    uint32_t partition,
+    vector_size_t size,
+    BufferPtr indices,
+    const RowVectorPtr& input) {
+  VELOX_DCHECK(spillEnabled());
+
+  if (isInputFromSpill()) {
+    spiller_->spill(partition, wrap(size, std::move(indices), input));
+  } else {
+    int spillChannel = 0;
+    for (const auto& channel : keyChannels_) {
+      spillChildVectors_[spillChannel++] = input->childAt(channel);
+    }
+    for (const auto& channel : dependentChannels_) {
+      spillChildVectors_[spillChannel++] = input->childAt(channel);
+    }
+    spiller_->spill(
+        partition,
+        wrap(
+            size,
+            std::move(indices),
+            tableType_,
+            spillChildVectors_,
+            input->pool()));
+  }
+}
+
+bool HashBuild::requestSpill(RowVectorPtr& input) {
+  VELOX_CHECK_GT(numSpillRows_, 0);
+  VELOX_CHECK_GT(numSpillBytes_, 0);
+
+  input_ = std::move(input);
+  if (spillGroup_->requestSpill(*this, future_)) {
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForSpill);
+    return false;
+  }
+  input = std::move(input_);
+  return true;
+}
+
+bool HashBuild::waitSpill(RowVectorPtr& input) {
+  if (!spillGroup_->needSpill()) {
+    return false;
+  }
+
+  input_ = std::move(input);
+  if (spillGroup_->waitSpill(*this, future_)) {
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForSpill);
+    return true;
+  }
+  input = std::move(input_);
```

Comment:
Why do we need to move input to input_, then back? Could we move input only if spillGroup_->waitSpill returns true?

Reply:
Yes. Thanks!